Migrate the underlying data store to a sorted set

erikogan opened the pull request:
Sorted sets are a more fully supported data structure than lists in Redis, and we can use the score of the sorted set to represent when they were added, and age them out so we don’t end up with a channel with a stack of 300+ entries.

Also add a `restack` command to bump yourself to the bottom. This was favored over silently doing this when a user already in the list stacks again, since that may lead to someone losing their place.

The collision error messages (as well as the “You have the floor” message) have been improved to remind folks to unstack when they are done, or restack if they wish to be added to the bottom.

Commit:
Migrate the underlying store to a sorted set so they can be timed out.

Semantic Version bumped to 1.0.0 for the API change.

Commit:
add stack[s] show as a valid list command

Since I used it in the spec, giving false positives

Commit:
Add the restack command

Commit:
Add upgrade path for existing data

Commit:
Reorganize code to make Metrics/AbcSize happy

Commit:
Fix the order of the collision predecessors

erikogan commented:
On further reflection, now that the messages have been improved to include everyone in front of you, I think it is less of an issue that a user might lose their place. Perhaps we should just automatically `restack` someone regardless if they are in the list or not.

… Especially since the annoyance of having to restack is going to happen _far_ more often than someone losing their place. Though I still worry about someone stacking energetically multiple times and losing their place.

edgeworthch commented on locales/en.yml:
```diff
@@ -4,9 +4,9 @@ en:
-          first: '%{user}, you have the floor!'
+            first: '%{user} already has the floor. Please use `unstack` when you are done, or `restack` to change your position to the bottom.'
+            after: '%{user} is already in the stack list, behind %{after}. Please use `restack` if you wish to change your position to the bottom.'
+          first: '%{user}, you have the floor! Please use `unstack` when you are done.'
```
I love these usability changes.  Agreed with your comment on the PR that `stack` should probably just be `restack` if a username is already in the queue.

edgeworthch commented on locales/en.yml:
```diff
@@ -32,16 +32,28 @@ en:
+        readd:
+          help:
+            command: 'restack [@user]'
+            description: 'remove and readd yourself (or @user) to the stack list'
```
Might be better to go with "re-add" since "readd" can be seen as a typo'd version of "read"

erikogan replied:
Agreed, but I think I’ve convinced myself to remove this command and have the `stack` command do it instead.

edgeworthch commented on spec/lita/handlers/stacker_spec.rb:
```diff
@@ -185,6 +204,76 @@
+    context 'when the message is in a room' do
+      include_context 'in a room'
+
+      it_behaves_like 'it clears old stacks'
```
Do you want to cover the transition case, where a single username might show up multiple times in the queue, and determine proper behavior there?  Or just go for a clean break instead?

edgeworthch replied:
ohh, switching to sorted sets means that all stacks would be cleared anyway, is that right?

erikogan replied:
The module to upgrade from lists to sorted sets [explicitly covers that case](https://github.com/change/lita-stacker/blob/71809d5b08698289ef36a68ba31fca3b4f500155/lib/lita/handlers/stacker/upgrade/sorted_sets.rb#L34) ([tested here](https://github.com/change/lita-stacker/blob/71809d5b08698289ef36a68ba31fca3b4f500155/spec/lita/handlers/stacker/upgrade/sorted_sets_spec.rb#L36-L39). Perhaps that spec should read “…maintaining the correct order?”) because, yes. zsets only allow one instance of a given lexical value.

erikogan replied:
…Which, I just realized, means that the new `stack` command can just add a user with a new score and the zset will handle everything else.

edgeworthch commented on lib/lita/handlers/stacker.rb:
```diff
@@ -1,9 +1,13 @@
   module Handlers
     class Stacker < Handler
-      VERSION = '0.2.0'
+      VERSION = '1.0.0'
```
the interface is backward-compatible, no?  This could also be considered a minor version change due to that, I would think

erikogan replied:
I almost added a comment on this line, but realized it would live past this moment-in-time change so I added it to the git comment (but not my reasoning):

It depends what you consider “the API,” since that’s what SemVer is concerned with. Yes, the user interface underwent a minor version update, but the underlying data interchange is not in any way backward compatible. …But I think I just talked myself out of it, because “your data{model,store,base} is not an API.”

edgeworthch replied:
Good point about the underlying data being lost with this upgrade.  You've convinced me that it could be considered a major change due to that.

edgeworthch reviewed (commented):
Only skimmed so far, but seems like a solid improvement.  I like the idea of replacing `restack` with simply allowing people to `stack` again to be moved to the end (with a corresponding explanatory text response)

Commit:
Revert "Add the restack command"

This reverts commit 9bb97744d85c85f167ecd602b064fe19f4204952.

Commit:
PR Feedback: Now that the stack feedback includes the whole list of
people it is less likely a user will lose their place. Instead of
forcing them to restack themselves, just do it for them if they stack
again.

While this _may_ cause folks to lose their place, it is offset by the
annoyance factor of having to restack when you did not realize you were
still on the list. Ultimately this was trying to solve a small risk
social problem with an invasive technical solution.

Commit:
add rubocop-rspec

Commit:
Reinstate minor version bump

A data{model,store,base} is not an API.

Commit:
Add an explicit test for announcing the whole stack

Commit:
PR Feedback: be clearer about what the dedup spec is testing

Commit:
Move the active_support require inside the method that uses it, so the
gemspec can be loaded without requiring active_support

Otherwise it is a circular dependency in bots where the gem is used. The
gemspec cannot be loaded without active_support, and the gemspec is
needed to assert the active_support dependency.

Commit:
Migrate to storing user IDs instead of mention names

So that operating on other users is more robust.
